jung-algorithms/src/main/java/edu/uci/ics/jung/layout/algorithms/FRBHVisitorLayoutAlgorithm.java

```diff
@@ -31,12 +31,12 @@
   @Override
   public void visit(LayoutModel<N> layoutModel) {
     super.visit(layoutModel);
-    tree = new BarnesHutQuadTree(layoutModel);
+    tree = new BarnesHutQuadTree(layoutModel.getWidth(), layoutModel.getHeight());
   }
 
   @Override
   public synchronized void step() {
-    tree.rebuild();
+    tree.rebuild(layoutModel.getGraph().nodes(), layoutModel);
```

Comment:
If you're passing in the LayoutModel anyway, why pass in the nodes separately here?

Unless I've misunderstood how you've changed the design, the LayoutModel is the source of truth for the graph anyway, so I don't think that you'd want to make it possible for someone to pass in a set of nodes that wasn't that of the graph the LayoutModel was retaining.

Reply:
because I don't want the BarnesHutQuadTree to know what a LayoutModel is. All is needs is a collection of nodes and a mapping to Points

Reply:
I also want to be able to unit test it without needing a LayoutModel (and then a graph)
It would be fine to create a BarnesHutQuadTree with a collection of things that are not even in a graph.

Reply:
I am passing in the layoutModel, you are correct, but the BHQT thinks that it is a Function<N,Point>

Reply:
Have a look at BarnesHutQuadTreeTests.java (new unit test class) and you will see that there is not a Graph nor a LayoutModel in sight.

Reply:
I understand that one can use a BarnesHutQuadTree outside the context of a graph.  I guess part of the question is whether we're providing a general-purpose BH implementation, or one for use with our layouts.  I had assumed the latter, in which case just using the LayoutModel seems like the appropriate choice to avoid the source-of-truth problem I mentioned earlier.

Reply:
I'm offering to contribute a BarnesHutQuadTree (and even an OctTree) that is easy to unit test in isolation from the effects of things like LayoutModel and Graph implementations, and seems to work well with 2 of the JUNG force directed layout algorithms. Why in the world would you choose to make it depend on the way Graph and LayoutModel are implemented just to ensure that it works as expected, and just to make the source code look cleaner?

Reply:
Would you like to have a subclass that takes a LayoutModel as the argument, and preserves the intent to tie together the nodes with their mapping, and still leaves the BHQT testable in isolation?

Reply:
My concern is as you said: you've spent a lot of time rearchitecting things to get us to the point where it's much easier to enforce that there's a single source of truth for the graph being visualized, and I don't want to compromise that success without a clear benefit.

Thanks for making the change.